### PR TITLE
add projection example for multiple aggregates

### PR DESCRIPTION
**Pull request by svalaskevicius**

### CODE REVIEW

**Comment by svalaskevicius** [comment]
@Fristi what do you think? can you spot anything that could be improved easily?

File: build.sbt [hunk 1]
```diff
@@ -6,13 +6,14 @@ scalaVersion := "2.11.7"
...
 libraryDependencies += "com.lihaoyi" %% "pprint" % "0.3.6"
+libraryDependencies += "com.lihaoyi" %% "upickle" % "0.3.6"
 libraryDependencies += "com.chuusai" %% "shapeless" % "2.2.5"
 
 resolvers += Resolver.sonatypeRepo("releases")
 
 addCompilerPlugin("org.spire-math" %% "kind-projector" % "0.6.3")
 
-wartremoverErrors ++= Warts.allBut(Wart.Any, Wart.Nothing, Wart.Throw, Wart.IsInstanceOf, Wart.AsInstanceOf, Wart.NoNeedForMonad)
+// wartremoverErrors ++= Warts.allBut(Wart.Any, Wart.Nothing, Wart.Throw, Wart.IsInstanceOf, Wart.AsInstanceOf, Wart.NoNeedForMonad)
```
Comment (svalaskevicius) [inline build.sbt]: TODO: this is removed as there are lots of errors when using upickle. Options so far:
- investigate how to ignore specifically upickle errors
- swap upickle with an alternative lib
- disable wart remover :(

File: src/main/scala/Cqrs/Projection.scala [hunk 2]
```diff
@@ -1,40 +1,16 @@
 package Cqrs
 
 import Cqrs.Aggregate._
-import InMemoryDb._
-
-import scala.collection.immutable.TreeMap
+import DbAdapters.InMemoryDb._
```
Comment (svalaskevicius) [inline src/main/scala/Cqrs/Projection.scala]: TODO, referencing the actual db adapter is still bad here - extract `consumeEvents`?

File: src/main/scala/Domain/Counter.scala [hunk 3]
```diff
@@ -4,22 +4,26 @@ import Cqrs._
 import Cqrs.Aggregate._
 import cats.data.{ Xor, XorT }
 import cats.syntax.flatMap._
+import Cqrs.DbAdapters.InMemoryDb.createEventDataConsumer
```
Comment (svalaskevicius) [inline src/main/scala/Domain/Counter.scala]: TODO

File: src/main/scala/lib/HList.scala [hunk 4]
```diff
@@ -1,87 +1,30 @@
...
-    implicit def iteratedApplyUpdate1[S, H, T <: HList, B](implicit iTail: ApplyUpdate1[S, T, B]): ApplyUpdate1[S, H :: T, B] =
-      new ApplyUpdate1[S, H :: T, B] {
-        def apply(hc: H :: T, f: S => (S, B)) = {
-          val r = iTail(hc.tail, f)
-          (hc.head :: r._1, r._2)
-        }
-      }
-  }
-  def applyUpdate1[S, B, T <: HList](hc: T, f: S => (S, B))(implicit ev: ApplyUpdate1[S, T, B]) = ev(hc, f)
+  def kMap[S[_], B[_], T <: HList, tOut <: HList](hc: T, f: S ~> B)(implicit ev: KMapper[S, T, B, tOut]): tOut = ev(hc, f)
```
Comment (svalaskevicius) [inline src/main/scala/lib/HList.scala:28]: hmm, it seems that the map from shapeless is not easily reusable as it generates the `Out` type via abstract type instead of type parameters. this causes the usage site be dependant on the operation, although it needs to be exactly the same.

this version is at least compatible and is not creating the hlist from scratch

**Comment by svalaskevicius** [comment]
Hi @Fristi, any feedback on this version? 

currently the projection can be replayed from the start (events ordered by their creation) regardless of where in the process was it added. 

there are still some items to refactor, esp direct usage of the `inMemoryDb` in projections

File: src/main/scala/Domain/Counter.scala [hunk 5]
```diff
@@ -52,21 +56,15 @@ object CounterProjection {
...
-    def handle(id: AggregateId, e: Event, d: Data) = e match {
-      case Created(id) =>
-        println("created " + id); d
-      case Incremented =>
-        println("+1"); d.updated(id, d.get(id).fold(1)(_ + 1))
-      case Decremented => println("-1"); d.updated(id, d.get(id).fold(-1)(_ - 1))
-    }
-  }
+  def emptyCounterProjection = Projection.empty[Data](TreeMap.empty, List(
+    (Counter.tag, createEventDataConsumer( (d: Data, t: Tag, id: AggregateId, v: Int, e: Counter.Event) => {
```
Comment (svalaskevicius) [inline src/main/scala/Domain/Counter.scala]: TODO: create intermediate type here

extract serialisation to a level above DB adapter operations and leave DB only for storing / retrieving strings?

File: build.sbt (hunk 1 above)
Comment (Fristi) [inline build.sbt]: Decouple encoding/decoding of events? So we let our users pick their own serialization lib

File: src/main/scala/Cqrs/Aggregate.scala [hunk 6]
```diff
@@ -23,9 +28,9 @@ object Aggregate {
   final case class VersionedEvents[E](version: Int, events: List[E])
 
   sealed trait EventDatabaseOp[E, A]
-  final case class ReadAggregateExistence[E](id: AggregateId) extends EventDatabaseOp[E, Error Xor Boolean]
-  final case class ReadAggregate[E](id: AggregateId, fromVersion: Int) extends EventDatabaseOp[E, Error Xor List[VersionedEvents[E]]]
-  final case class AppendAggregateEvents[E](id: AggregateId, events: VersionedEvents[E]) extends EventDatabaseOp[E, Error Xor Unit]
+  final case class ReadAggregateExistence[E](tag: Tag, id: AggregateId) extends EventDatabaseOp[E, Error Xor Boolean]
```
Comment (Fristi) [inline src/main/scala/Cqrs/Aggregate.scala:31]: In akka-persistence events get a `persistenceId` (which is a `aggregateId`) and zero or more tags. In these cases you have to supply both. It would be nice to read events by a specific tag or a aggregateId. When writing you should write them along with one aggregatedId, but zero or more tags. Checking if a aggregate exists should only be done with a `aggregateId` I think?

File: src/main/scala/Cqrs/BatchRunner.scala [hunk 7]
```diff
@@ -4,72 +4,44 @@ import cats.data.Xor
...
 import Cqrs.Aggregate.{ AggregateDef, AggregateState, EventDatabaseWithFailure, Error }
-import Cqrs.InMemoryDb.{ DbBackend, runInMemoryDb }
-import lib.{ HList, HNil }
-import lib.HList._
+import Cqrs.DbAdapters.InMemoryDb.{EventSerialisation, DbBackend, runInMemoryDb}
+import cats.~>
 
-object BatchRunner {
-  def empty = BatchRunner[HNil, HNil.type](HNil, HNil)
+import shapeless._
```
Comment (Fristi) [inline src/main/scala/Cqrs/BatchRunner.scala:10]: Nice :+1:

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+import cats.std.all._
+
+import scala.collection.immutable.TreeMap
+
+
+object InMemoryDb {
+
+  trait EventSerialisation[E] {
+    def serialise(d: E): String
+    def unserialise(s: String): E
```
Comment (Fristi) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: Encode/decode might be better names here. Decoding might fail, so use `Option` or `Either` here I guess

**Comment by Fristi** [comment]
Added some comments here and there. 

Overall I need to take a closer look what all constructs are doing conceptually.

File: build.sbt (hunk 1 above)
Comment (svalaskevicius) [inline build.sbt]: Yep, that should be already possible :) 

I wonder if the wart remover was a good idea to have.. It works very well for the code in this repo but also limits the libraries we can use and a couldn't find a good way to disable it for implicit lookups from ext libs...

File: src/main/scala/Cqrs/Aggregate.scala (hunk 6 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/Aggregate.scala:31]: The semantics of the tags in this PR are a bit different from akka - its more like aggregate class - eg all counters, or all doors. Checking if exists also needs it in this case - "is there a counter (by tag) identified by this id". 

Projections tell what classes / tags of aggregates they are interested in and receive all aggregates within them

File: src/main/scala/Cqrs/BatchRunner.scala (hunk 7 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/BatchRunner.scala:10]: Still had to leave my hlist with the kmap operation - looks like shapeless has output type specified by operations and couldn't find an easy way to implement this kind of mapping (to the exactly same type, specified by caller). At least it's using shapeless type for the hlist data itself :)

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala (hunk 8 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: Oh yes, there is another similar place I left as todo (will solve before this is merged) 

Thanks!

File: src/main/scala/Cqrs/Projection.scala (hunk 2 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/Projection.scala]: solved

File: src/main/scala/Domain/Counter.scala (hunk 3 above)
Comment (svalaskevicius) [inline src/main/scala/Domain/Counter.scala]: solved

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala (hunk 8 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: done :)

File: src/main/scala/Eventflow.scala [hunk 9]
```diff
@@ -82,7 +80,12 @@ object Eventflow {
         fold(err => { println("Error occurred: " + err._2); err._1 }, r => { println("OK"); r._1 })
 
       @SuppressWarnings(Array("org.brianmckenna.wartremover.warts.OptionPartial", "org.brianmckenna.wartremover.warts.ExplicitImplicitTypes"))
-      val _ = pprint.pprintln(db_, colors = pprint.Colors.Colored)
+      val _ignore1 = pprint.pprintln(runner1, colors = pprint.Colors.Colored)
```
Comment (svalaskevicius) [inline src/main/scala/Eventflow.scala]: interesting that after some commit in this PR the pprint lost colours :/ (now its just showing toString result)

File: src/main/scala/Cqrs/Projection.scala [hunk 10]
```diff
@@ -1,40 +1,19 @@
...
+final case class Projection[D](lastReadOperation: Int, data: D, dbConsumers: List[(Tag, EventDataConsumer[D])]) {
 
-  import Projection._
-
-  def applyNewEventsFromDb[E](db: DbBackend[E])(implicit handler: Handler[E, D]): Projection[D] =
-    applyNewEventsFromDbToProjection(db, this)
+  def applyNewEventsFromDb[Db: Backend](db: Db): Projection[D] = {
+    val updatedProjectionData = Database.consumeDbEvents(db, lastReadOperation, data, dbConsumers)
+    updatedProjectionData.fold({
+      println("failure to update projection")
```
Comment (svalaskevicius) [inline src/main/scala/Cqrs/Projection.scala]: todo: move error handling up

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      }
+    }
+
+  implicit def dbBackend: Backend[DbBackend] = new Backend[DbBackend] {
+    def runDb[E: EventSerialisation, A](database: DbBackend, actions: EventDatabaseWithFailure[E, A]): Error Xor (DbBackend, A) = {
+      val (db, r) = actions.value.foldMap[Db](transformDbOpToDbState).run(database).run
+      r map ((db, _))
+    }
+
+    def consumeDbEvents[D](database: DbBackend, fromOperation: Int, initData: D, query: EventDataConsumerQuery[D]): Option[(Int, D)] = {
```
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: todo: replace option with xor

**Comment by svalaskevicius** [comment]
ok, looks a bit better now I hope :)

@Fristi can you spot anything that has to be fixed before merging?

File: build.sbt [hunk 12]
```diff
@@ -6,14 +6,13 @@ scalaVersion := "2.11.7"
...
 libraryDependencies += "org.spire-math" %% "cats" % "0.2.0"
 libraryDependencies += "com.lihaoyi" %% "pprint" % "0.3.6"
+libraryDependencies += "com.lihaoyi" %% "upickle" % "0.3.6"
 libraryDependencies += "com.chuusai" %% "shapeless" % "2.2.5"
 
 resolvers += Resolver.sonatypeRepo("releases")
 
 addCompilerPlugin("org.spire-math" %% "kind-projector" % "0.6.3")
 
-wartremoverErrors ++= Warts.allBut(Wart.Any, Wart.Nothing, Wart.Throw, Wart.IsInstanceOf, Wart.AsInstanceOf, Wart.NoNeedForMonad)
```
Comment (Fristi) [inline build.sbt:15]: No more wartremover? :(

File: src/main/scala/Cqrs/Aggregate.scala (hunk 6 above)
Comment (Fristi) [inline src/main/scala/Cqrs/Aggregate.scala:31]: I see, makes abit sense. However, how would you read all events of a counter for example without knowing all the aggregateIds? `case class ReadAggregate[E](tag: Tag, id: AggregateId, fromVersion: Int)` requires you to specify a AggregateId

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    lastOperationNr: Int
+  )
+  type Db[A] = State[DbBackend, A]
+
+  def newInMemoryDb: DbBackend = DbBackend(TreeMap.empty, TreeMap.empty, 0)
+
+  def readFromDb[E: EventSerialisation](database: DbBackend, tag: Tag, id: AggregateId, fromVersion: Int): Error Xor List[VersionedEvents[E]] = {
+
+    def getById(id: AggregateId)(t: TreeMap[String, TreeMap[Int, List[String]]]) = t.get(id.v)
+    def unserialise(d: String) = implicitly[EventSerialisation[E]].decode(d)
```
Comment (Fristi) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: Would prefer decode here

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    def unserialiseEvents(d: List[String])(implicit t: Traverse[List]): Error Xor List[E] = t.sequence[Xor[Error, ?], E](d map unserialise)
+    def dbRecordToMaybeVersionedEvent(dbrec: (Int, List[String])) = {
```
Comment (Fristi) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: It is unclear to me what the intent of the function is. 

There are two hard things here:
1. The name doesn't describe it for me.
2. The (tuples) types don't describe it, it would be good to use case classes here ( but you would need to come up with good names again :) ).

File: build.sbt (hunk 12 above)
Comment (svalaskevicius) [inline build.sbt:15]: I like wartremover.. when it works, but it seems to restrict libraries we can use conveniently at this point.. e.g. uPickle fails quite dramatically and I couldnt find a way to selectively disable it (as it fails when looking for implicits).. 

do you have any ideas how can I restore it? (should I use another serialisation library?)

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala (hunk 14 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: hmm, let me try again then :)

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala (hunk 13 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: will change :)

File: src/main/scala/lib/foldM.scala [hunk 15]
```diff
@@ -0,0 +1,13 @@
+package lib
+
+import cats.Monad
+
+import scala.language.higherKinds
+
+object foldM {
```
Comment (Fristi) [inline src/main/scala/lib/foldM.scala:7]: Isn't this already in cats somewhere? :)

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala (hunk 13 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: done, thanks :)

File: src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala (hunk 14 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/DbAdapters/InMemoryDb.scala]: done, hopefully clearer - added the return type as well :)

File: src/main/scala/lib/foldM.scala (hunk 15 above)
Comment (svalaskevicius) [inline src/main/scala/lib/foldM.scala:7]: sadly no... there is an open PR for it - https://github.com/non/cats/pull/356 but its on hold

File: src/main/scala/lib/HList.scala (hunk 4 above)
Comment (Fristi) [inline src/main/scala/lib/HList.scala:28]: You can solve that by introducing a type class (you already did) and use a `Aux` type on the companion object.

This is how most shapeless operations are implemented see for example:

https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/ops/hlists.scala#L1721

Prepend here defines a type Out which can be parameterized by using the Aux in the companion object.

**Comment by Fristi** [comment]
Overall I think it's nice to get projections out there! There are some things we could improve here :) Sometimes I find it not easy to digest some of you code. How could we improve on that?
I think with a few modifications we can clarify it a lot, what do you think?
- Private/public or local functions. If a function does not need public scope, make it local (if you need it for recursive loop or something) or make it a private one if you use it multiple times. This makes modules a bit easier to digest.
- Don't use tuples in public API's, instead use case classes. They can describe what the values in the tuples are by a good name. I essence case classes are tuples, but with names :)
- Use descriptive parameter names in case classes and functions. In the example I saw you use parameters names like `k`, does not really reveal what the intent of parameter is
- Rather use (implicit e: Backend[T]) for implicits if you use it multiple times. Makes the code more concise (and easier to read), since you don't have to write implictly each time.
- For concepts like Backend or EventDataConsumerQuery it doesn't reveal it's intention immediately. Would be nice to introduce some docs maybe? 
- Group concepts and put them in packages or separate files. 

I can help you out with refactoring some parts! For now it would be good to merge this :+1:

File: src/main/scala/lib/foldM.scala (hunk 15 above)
Comment (Fristi) [inline src/main/scala/lib/foldM.scala:7]: Ah cool, let's address that later then!

File: src/main/scala/lib/HList.scala (hunk 4 above)
Comment (svalaskevicius) [inline src/main/scala/lib/HList.scala:28]: I've tried adding Aux in the same way however that results in the same error.. have I missed something? In general reasoning Aux is kind of similar as the default shapeless mapper...

```
Error:(24, 180) type mismatch;
 found   : Cqrs.BatchRunner[Db,shapeless.::[Cqrs.Projection[D],m.Out]]
 required: Cqrs.BatchRunner[Db,shapeless.::[Cqrs.Projection[D],PROJS]]
Error occurred in an application involving default arguments.
  def addProjection[D](proj: Projection[D])(implicit m: KMapper[Projection, PROJS, Projection]): BatchRunner[Db, Projection[D] :: PROJS] = copy(projections = proj :: projections).runProjections
                                                                                                                                                                                   ^
```

File: src/main/scala/Cqrs/Aggregate.scala (hunk 6 above)
Comment (svalaskevicius) [inline src/main/scala/Cqrs/Aggregate.scala:31]: for this, there is currently a separate Database method as defined by the typeclass - https://github.com/svalaskevicius/eventflow/blob/multiEventsProjection/src/main/scala/Cqrs/Database.scala#L12

its not available in the aggregate actions as no aggregate should ever need that.. I think :)

**Comment by svalaskevicius** [comment]
completely agree! :) in fact, I thought I am already doing most of those items :D  apart from short names and no docs. 

happy to update the missed parts - in the end, I will have to read this as well even after I forget how it was implemented :)

will have a skim over again and see what have I missed, thanks!

**Comment by svalaskevicius** [comment]
@Fristi are these places that you meant - https://github.com/svalaskevicius/eventflow/commit/8a34878c930644b01866e42c03d894c008300d95 ? happy to improve more if you see something obvious I missed again :)

will merge - but we can always improve this in other PRs (and I hope we will :) )

and thanks for the review!
